### DISCUSSION TITLE
How best to use javascript hooks in page transitions?

### DISCUSSION THREAD

**Discussion by chasebank** [nuxt/nuxt#16526] [Questions]
Nuxt 3 seems great, but I'm back to some confusion/friction I've always had with Nuxt - page/layout transition hooks.

If I have an element **outside** the page/route component (a static navigation for example, shared across all pages/layouts), and I want to animate it **before** the new route enters, and then again **after** the new route enters, how can this be done?

My solution in v2 was throw away Nuxt's page transitions and layouts (unfortunate - because those are key features), and use only the single default layout as a pseudo "app" component, so I can wrap the `<nuxt/>` element in a normal Vue `<transition>` and use _real_ hooks. Not ideal, but worked.

Having access to `app.vue` in Nuxt 3 is progress, but I'm still not sure how to leverage page transitions.

`beforeUpdate` and `updated` in a directive seems like it could work, but also not sure how that translates to Nuxt.

I see [this](https://v3.nuxtjs.org/getting-started/transitions#javascript-hooks):
```javascript
definePageMeta({
  pageTransition: {
    onBeforeEnter: (el) => {
      console.log('Before enter...')
    },
    onEnter: (el, done) => {},
    onAfterEnter: (el) => {}
  }
})
```
But, wouldn't that have to be included in every page? I could standardize using a "page" component to avoid repetition, but then I'd have to solve for affecting elements outside that component (in `app.vue`), and still seems generally strange.

Aren't I trying to hook into the **transition** component? Not the transition**ed** element? Shouldn't these hooks be handled in the parent of the transition component? (which in this case I'd assume to be layout?)

**Comment by ghost** [comment]
You could ould target other elements via CSS in some sort of global styling (app.vue or nuxt.config.ts css-option), if you use the name option:

Applies to name="page"

```
.page-enter-active,
.page-leave-active 
page-enter-active ~ .footer ,
.page-leave-active ~ .footer{
  transition: opacity 0.4s;
}
.page-enter-from,
.page-leave-to,
.page-enter-from  ~ .footer,
.page-leave-to ~ .footer {
  opacity: 0;
}
```

~ is a general sibling selector, + is a sequential selector (will only apply if it's the direct next element)

With that no additional js necessary. I know CSS-selecting can be a bit tricky that way, but it should be more performant than extra js dom manipulation.

**Reply by chasebank** [reply]
Thanks for that suggestion, but as you point out this would only target direct siblings of the page component, inside the layout. Which would indeed be the case with how I used to do things - but now that Nuxt3 offers an `app.vue` component, ideally I'm trying to avoid using layout for app components.

Aside from that, I need to leverage javascript hooks to affect things *before* the new page component mounts and affects layout. (transition mode would also not be a sufficient alternative)
